src/transformers/models/deprecated/open_llama/modeling_open_llama.py

```diff
@@ -124,15 +124,32 @@ def _set_cos_sin_cache(self, seq_len, device, dtype):
         self.register_buffer("cos_cached", emb.cos()[None, None, :, :].to(dtype), persistent=False)
         self.register_buffer("sin_cached", emb.sin()[None, None, :, :].to(dtype), persistent=False)
 
-    def forward(self, x, seq_len=None):
-        # x: [bs, num_attention_heads, seq_len, head_size]
+    def rotate_half(self, x):
+        """Rotates half the hidden dims of the input."""
+        x1 = x[..., : x.shape[-1] // 2]
+        x2 = x[..., x.shape[-1] // 2 :]
+        return torch.cat((-x2, x1), dim=-1)
+
+    def apply_rotary_pos_emb(self, q, k, cos, sin, position_ids):
+        # The first two dimensions of cos and sin are always 1, so we can `squeeze` them.
+        cos = cos.squeeze(1).squeeze(0)  # [seq_len, dim]
+        sin = sin.squeeze(1).squeeze(0)  # [seq_len, dim]
+        cos = cos[position_ids].unsqueeze(1)  # [bs, 1, seq_len, dim]
+        sin = sin[position_ids].unsqueeze(1)  # [bs, 1, seq_len, dim]
+        q_embed = (q * cos) + (self.rotate_half(q) * sin)
+        k_embed = (k * cos) + (self.rotate_half(k) * sin)
+        return q_embed, k_embed
+
+    def forward(self, q, k, position_ids, seq_len=None):
```

Comment:
changing the forward pass this way is breaking 😢 
We should try to avoid that!

Reply:
If the forward pass was left as `def forward(self, x, seq_len=None)`, then the below lines are called when they don't have to be.
https://github.com/huggingface/transformers/blob/75e9b8d69b7a0a4c4c8d43430c83dd2bd0249e8c/src/transformers/models/deprecated/open_llama/modeling_open_llama.py#L145-L151
https://github.com/huggingface/transformers/blob/75e9b8d69b7a0a4c4c8d43430c83dd2bd0249e8c/src/transformers/models/deprecated/open_llama/modeling_open_llama.py#L135-L138

Reply:
If the forward pass was left as `def forward(self, x, seq_len=None)`, then the below lines are called twice when they don't have to be.
https://github.com/huggingface/transformers/blob/75e9b8d69b7a0a4c4c8d43430c83dd2bd0249e8c/src/transformers/models/deprecated/open_llama/modeling_open_llama.py#L145-L151
https://github.com/huggingface/transformers/blob/75e9b8d69b7a0a4c4c8d43430c83dd2bd0249e8c/src/transformers/models/deprecated/open_llama/modeling_open_llama.py#L135-L138

Reply:
If the forward pass was changed to `def forward(self, x, seq_len=None, pos_ids=None)` to maintain backward compatibility, then the below lines are called twice when they don't have to be.
https://github.com/huggingface/transformers/blob/75e9b8d69b7a0a4c4c8d43430c83dd2bd0249e8c/src/transformers/models/deprecated/open_llama/modeling_open_llama.py#L145-L151
https://github.com/huggingface/transformers/blob/75e9b8d69b7a0a4c4c8d43430c83dd2bd0249e8c/src/transformers/models/deprecated/open_llama/modeling_open_llama.py#L135-L138